src/snarky/parties-helpers.ts

```diff
@@ -297,46 +229,122 @@ function fromFieldsReversed(
   return Leaves.fromFields(typeData.type, fields, aux);
 }
 
-function check(typeData: Layout, value: any, customTypes: CustomTypes): void {
+function check(typeData: Layout, value: any, customTypes: CustomTypes) {
+  return mapReduce<any, void>(
+    {
+      map(type, value) {
+        return Leaves.check(type, value);
+      },
+      mapCustom(type, value) {
+        return type.check(value);
+      },
+      reduceArray() {},
+      reduceObject() {},
+      reduceFlaggedOption() {},
+      reduceOrUndefined() {},
+      customTypes,
+    },
+    typeData,
+    value
+  );
+}
+
+function toInput(typeData: Layout, value: any, customTypes: CustomTypes) {
+  return mapReduce<any, Leaves.Input>(
+    {
+      map(type, value) {
+        return Leaves.toInput(type, value);
+      },
+      mapCustom(type, value) {
+        return type.toInput(value);
+      },
+      reduceArray(array) {
+        let acc: Leaves.Input = { fields: [], packed: [] };
+        for (let { fields, packed } of array) {
+          if (fields) acc.fields!.push(...fields);
+          if (packed) acc.packed!.push(...packed);
+        }
+        return acc;
+      },
+      reduceObject(keys, object) {
+        let acc: Leaves.Input = { fields: [], packed: [] };
+        for (let key of keys) {
+          let { fields, packed } = object[key];
+          if (fields) acc.fields!.push(...fields);
+          if (packed) acc.packed!.push(...packed);
+        }
+        return acc;
+      },
+      reduceFlaggedOption({ isSome, value }) {
+        return {
+          fields: value.fields,
+          packed: isSome.packed!.concat(value.packed ?? []),
+        };
+      },
+      reduceOrUndefined(_) {
+        return {};
+      },
+      customTypes,
+    },
+    typeData,
+    value
+  );
+}
+
+type MapReduceSpec<T, R> = {
```

Comment:
I think this is semantically more of a `fold` ala `Fields.fold` in the ocaml code that’s generated from `deriving fields`. I think something like FoldSpec and genericFold or layoutFold?